### ISSUE TITLE
How to get rid of quick flickering of untranslated content

### ISSUE THREAD

**Issue by vlasenko** [angular-translate/angular-translate#1311]
Right now if you point debugger to translate-cloak directive implementation into onReady block and refresh the page, that uses the
```
<h1 translate>some.key<h1/>
```
directives you will still be able to see untranslated keys. How to get rid of this? I would like to cloak these keys until their translation promises will be actually resolved and not until translation engine "will be ready to translate content" as it is made now. I don't want to show untranslated keys to user even during a millisecond.

**Comment by toxidev** [comment]
Hello!
I'm not an expert, but I think you can try this <h1 translate="some.key"><h1/>

**Comment by knalli** [comment]
@vlasenko  I have not understand your issue completely.

`translate-cloak` only controls the existince of a css class (on the specified element).. you still have to add it. Alternatively, you can use `<h1 translate="some.key"></h1>` which will render only AFTER the first resolving (so no FOUC)

**Comment by vlasenko** [comment]
@knali I want to prevent flow of untranslated content. Translate-cloak DOES NOT completely prevents this, because it doesn't wait until all translation futures on a page get resolved. I have described in details in my starting post how to reproduce this. What exactly do you not understand?

**Comment by knalli** [comment]
@vlasenko That would mean that some translations will be resolved after the translations are being loaded and available. That sounds strange.

Anway: What about specifying an additional translationId within translate-cloak? Does this work for you?

**Comment by vlasenko** [comment]
> Anway: What about specifying an additional translationId within translate-cloak? Does this work for you?

No, I still see untranslated keys after the finish of first $digest cycle.

> Alternatively, you can use <h1 translate="some.key"></h1> which will render only AFTER the first resolving (so no FOUC)

This somewhat works, but I don't see such elements in the DOM after the end of first digest, so I suspect there will be some page element jumping issue when they appear, but I'm not sure.

**Comment by vlasenko** [comment]
Again what I want is FOUC issue prevented completely, e.g. I don't show the region of page protected by say translate-cloak until it's fully ready.

**Comment by knalli** [comment]
@vlasenko  Okay, it looks like the cloak is broken when used async and sync translation at the same time.

http://plnkr.co/edit/9hyXezWDlFKZW8l9GDxu?p=preview I guess you are using both static translations as well as dynamic loaded (via a loader). If you uncomment the ll.19-24, it won't work as expected any more.

**Comment by vlasenko** [comment]
Aha, thanks for the pointer! So the workaround for me is to completely switch to static translations for now?

**Comment by knalli** [comment]
Ehm.. well, the example shows it is working only using the loader (async way). Actually you don't need a cloak when using static translations! Because this FOUC will being handled by ng-cloak already. translate-cloak is only made for the time between AJS is initialized (and ng-cloak will be removed) and the first translation is available.

**Comment by vlasenko** [comment]
Yup, understood, thanks.

**Comment by knalli** [comment]
@vlasenko  Can you play with  http://plnkr.co/edit/tF8QqyMukl4XKinixKZP?p=info ? This is a 2.9 pre release containing fixes.

**Comment by vlasenko** [comment]
@knalli Thank you. I'm going to play with this version during next 2 days and will let you know how it goes.

**Comment by vlasenko** [comment]
@knalli It doesn't solve my issue, though it might work better then before. My issue is I want to prevent completely even for a nanosecond time showing of untranslated keys by web browser. To reproduce this in JavaScript debugger set the breakpoint at the end of removeCloak function in angular-translate and at the end of angular.js $digest method, refresh a page that contains async translation keys and when debugger stops in removeCloak press continue, when debugger stops in the end of $digest method you should see FOUC. This is not expected behavior from angular-translate because translate-cloak is supposed to prevent FOUC.

The plunkr provided does not completely simulate the described problem.

**Comment by knalli** [comment]
What is the difference between my plnkr and your situation? Perhaps fork it and make the appropriate change..

**Comment by vlasenko** [comment]
Why to reproduce it using plnkr at all? The steps to reproduce involve JavaScript debugger. It's easy to reproduce it this way.

**Comment by knalli** [comment]
Clearly it does not look like it would be so easy, since my plnkr seems to be unaffected and you cannot show me a demo. I'm interested in the *difference*. 

One note: Do not trust timings when debugging is enabled: it is heisenberg involved...

**Comment by vlasenko** [comment]
Well, I can show you a demo in a plunkr, but reproduction will still involve JavaScript debugger. If it will suffice you, then I will prepare it. I understand that you want to get rid of JavaScript debugger to work with this issue, it will be much more convenient to you, but this will require some time from me to investigate deeply angular and angular-translate to clear understand what's going on wrong and I have no time right now to do this and the issue is not so important for me. It will be nice to have it sorted out but without major time investments.

I disagree about timings. The thing is if translate-cloak is working properly there should be no moment when browser shows untranslated keys, but right now there is clearly such a moment _at the end of first angular digest cycle, when cloak is removed_.

**Comment by knalli** [comment]
If you are using `<h1 translate>some.key<h1/>` in your code, the text `some.key` will be rendered out. That's HTML. If you avoid this, you have to use the alternative way `<h1 translate="some.key"><h1/>` which will never write out the key (unless you provide such a error handler configuration).  http://plnkr.co/edit/NuE2EcEpY6yJ19fieJnY?p=preview

Is this the difference you are talking about?

**Comment by vlasenko** [comment]
Yes

**Comment by knalli** [comment]
Okay.. then the solution will be using `<elem translate="ID"></elem>`. There will be always an amount of time between html rendering and AJS (and then angular-translate directive) picking up and creating an instance and changing the DOM.

The background is simple: AJS is calling us (using the directive) on the already existing DOM.

I think we are done here?

**Comment by vlasenko** [comment]
Yes, I think so too. Thank you!

**Comment by knalli** [comment]
Thank you also!
